pkg/chrootarchive/archive.go

```diff
@@ -82,6 +82,17 @@ func untarHandler(tarArchive io.Reader, dest string, options *archive.TarOptions
 			return err
 		}
 	}
+	relDest, err := filepath.Rel(root, dest)
+	if err != nil {
+		return err
+	}
+	if relDest == "." {
+		relDest = "/"
+	}
+	if relDest[0] != '/' {
+		relDest = "/" + relDest
+	}
+	dest = relDest
```

Comment:
I didn’t try this and I might well be missing something, but how does this work on non-linux where `root`/`rootfd` are completely ignored?

Reply:
I don't think we should continue the charade of supporting containers/storage on non linux/freeBSD machines.

Reply:
I vaguely remember the tar/untar functions being used for non-graph-driver content as well. E.g. Podman’s `pkg/machine/ocipull/oci.go`.

Reply:
… and if we _were_ to drop support for other systems, I think the code would _have_ to be modified to fail to compile, rather than being allowed to write to unexpected destinations by neglect.

Reply:
> I didn’t try this and I might well be missing something, but how does this work on non-Linux where root/rootfd are completely ignored?

It doesn't.

> I vaguely remember the tar/untar functions being used for non-graph-driver content as well. E.g. Podman’s pkg/machine/ocipull/oci.go.

Hmm, good point, but in glancing slightly more deeply it seems like that...ah yes, notice the `archive` vs `chrootarchive`, so it's not using this code I think.

Of course, crossing with other threads you know of, I don't see why the podman-machine disk images are actually stored as container images (including being wrapped in a tarball) that we manually unpack vs directly as an OCI artifact which we fetch and stream directly to disk without an extra copy, which would be inherently more secure too. (Of course, if the podman machine archive today is malicious, you're kind of boned anyways because that machine ends up with write access to your home directory)

Reply:
> on non linux/freeBSD

Of course already needing to abstract over those two is pretty nontrivial...Linux has `openat2` but FreeBSD ended up doing things differently with `O_RESOLVE_BENEATH` etc. The current zstd:chunked code is hardcoded to Linux.

Since it's been at least a whole day since I've mentioned Rust here...in bootc/ostree we've been using https://github.com/bytecodealliance/cap-std/ and it's quite nice and elegant and well designed, and has [backends for both linux and freebsd](https://github.com/bytecodealliance/cap-std/tree/main/cap-primitives/src/rustix). In theory we could fork off a helper process using that to handle untar...it even has a whole "userspace fs resolver fallback" path that is what's used today on MacOS.

Reply:
> > I vaguely remember the tar/untar functions being used for non-graph-driver content as well. E.g. Podman’s pkg/machine/ocipull/oci.go.
> 
> Hmm, good point, but in glancing slightly more deeply it seems like that...ah yes, notice the `archive` vs `chrootarchive`, so it's not using this code I think.

Thanks, that’s an important nuance (while we do need to _partially_ continue to support c/storage on the other systems).

---

A longer comment in https://github.com/containers/storage/pull/2051#issuecomment-2260940574 , but basically I’m leaning towards the though that a Linux-only code path is not an improvement.

The untag code is _supposed_ to be secure in any environment; AFAICT the chroot is an _additional_ isolation, not something the callers specifically require because they depend on some chroot-only property.

> In theory we could fork off a helper process using that to handle untar...

The way `pkg/reexec` works here, there are no external helper binaries. In the post-RPM world of people just copying binaries around, that makes deployment easier — probably not for Podman and all its dependencies, but for some of the other lighter-weight callers.

Reply:
> I don't see why the podman-machine disk images are actually stored as container images

At least for the `pkg/machine/compression` code, that’s not necessarily the case — it’s a very generic decompressor handling things like ZIP, not compatible with traditional container images anyway.